### PR TITLE
optimize insecure tls cipher checks within one loop

### PR DESCRIPTION
**Pull request by KielChan**
**What type of PR is this?**

> /kind cleanup

**What this PR does / why we need it**:

Old insecure cipher are checked by two for loops nested and one extra loop to fetch all insecure cipher Info, [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L1014)
The new one will collect insecure ciphers within one loop to check all ciphers in flag valid or not.
```golang
func TLSCipherSuites(cipherNames []string) ([]uint16, []string, error) {
	if len(cipherNames) == 0 {
		return nil, nil, nil
	}
	ciphersIntSlice := make([]uint16, 0)
	insecureCipherNamesMatched := make([]string, 0)
	possibleCiphers := allCiphers()
	for _, cipher := range cipherNames {
		intValue, ok := possibleCiphers[cipher]
		if !ok {
			return nil, nil, fmt.Errorf("Cipher suite %s not supported or doesn't exist", cipher)
		}
		ciphersIntSlice = append(ciphersIntSlice, intValue)

		if _, ok := insecureCiphers[cipher]; ok {
			insecureCipherNamesMatched = append(insecureCipherNamesMatched, cipher)
		}
	}
	return ciphersIntSlice, insecureCipherNamesMatched, nil
}
```

Then:
```golang
//...
	tlsCipherSuites, insecureCipherNames, err := cliflag.TLSCipherSuites(kc.TLSCipherSuites)
	if err != nil {
		return nil, err
	}

	for _, cipherName := range insecureCipherNames {
		klog.Warningf("Use of insecure cipher '%s' detected.", cipherName)
	}

//...
```

Maybe that is cool than before.

**Which issue(s) this PR fixes**: 
none

Fixes # 
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by enj** [comment]
I would prefer to leave `TLSCipherSuites` as-is instead of trying to make it to do multiple things.  An extra `for` loop does not hurt anything.  Maybe a `WarnIfInsecureTLSCipherUsed` function makes sense to help with consistency.

**Comment by liggitt** [comment]
> I would prefer to leave `TLSCipherSuites` as-is instead of trying to make it to do multiple things. An extra `for` loop does not hurt anything.

Agreed.

**Comment by KielChan** [comment]
Fine, I will add another function to handle insecure cipher suites check, like `WarnIfInsecureTLSCipherUsed `, to make it clean.

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go [hunk 1]
```diff
@@ -109,21 +109,27 @@ func TLSCipherPossibleValues() []string {
...
+		return nil, nil, nil
 	}
 	ciphersIntSlice := make([]uint16, 0)
+	insecureCipherNamesMatched := make([]string, 0)
 	possibleCiphers := allCiphers()
 	for _, cipher := range cipherNames {
 		intValue, ok := possibleCiphers[cipher]
 		if !ok {
-			return nil, fmt.Errorf("Cipher suite %s not supported or doesn't exist", cipher)
+			return nil, nil, fmt.Errorf("Cipher suite %s not supported or doesn't exist", cipher)
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:124]:
```suggestion
			return nil, nil, fmt.Errorf("cipher suite %s not supported or doesn't exist", cipher)
```

see https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: cmd/kubelet/app/server.go [hunk 2]
```diff
@@ -1006,20 +1005,13 @@ func InitializeTLS(kf *options.KubeletFlags, kc *kubeletconfiginternal.KubeletCo
...
 	}
 
-	tlsCipherSuites, err := cliflag.TLSCipherSuites(kc.TLSCipherSuites)
+	tlsCipherSuites, insecureCipherNames, err := cliflag.TLSCipherSuites(kc.TLSCipherSuites)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(tlsCipherSuites) > 0 {
-		insecureCiphers := flag.InsecureTLSCiphers()
```
Comment (SergeyKanzhelev) [inline cmd/kubelet/app/server.go:1015]: is the method `InsecureTLSCiphers` ever used after this PR? Should it be removed?

**Review by SergeyKanzhelev — APPROVED** [review]

**Comment by SergeyKanzhelev** [comment]
> I would prefer to leave `TLSCipherSuites` as-is instead of trying to make it to do multiple things. An extra `for` loop does not hurt anything. Maybe a `WarnIfInsecureTLSCipherUsed` function makes sense to help with consistency.

I feel like the latest changes make sense

**Comment by SergeyKanzhelev** [comment]
seems to be legit build failure:

```
diff -Naupr -x _output /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/BUILD /tmp/verify-bazel.jN9xZB/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/BUILD
--- /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/BUILD	2020-07-08 20:41:36.078616374 +0000
+++ /tmp/verify-bazel.jN9xZB/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/BUILD	2020-07-08 20:42:49.533090460 +0000
@@ -119,7 +119,6 @@ go_library(
         "//staging/src/k8s.io/apiserver/pkg/util/openapi:go_default_library",
         "//staging/src/k8s.io/client-go/informers:go_default_library",
         "//staging/src/k8s.io/client-go/rest:go_default_library",
-        "//staging/src/k8s.io/component-base/cli/flag:go_default_library",
         "//staging/src/k8s.io/component-base/logs:go_default_library",
         "//vendor/github.com/coreos/go-systemd/daemon:go_default_library",
         "//vendor/github.com/emicklei/go-restful:go_default_library",
diff -Naupr -x _output /home/prow/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/BUILD /tmp/verify-bazel.jN9xZB/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/BUILD
--- /home/prow/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/BUILD	2020-07-08 20:41:36.078616374 +0000
+++ /tmp/verify-bazel.jN9xZB/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/BUILD	2020-07-08 20:42:49.533090460 +0000
@@ -119,7 +119,6 @@ go_library(
         "//staging/src/k8s.io/apiserver/pkg/util/openapi:go_default_library",
         "//staging/src/k8s.io/client-go/informers:go_default_library",
         "//staging/src/k8s.io/client-go/rest:go_default_library",
-        "//staging/src/k8s.io/component-base/cli/flag:go_default_library",
         "//staging/src/k8s.io/component-base/logs:go_default_library",
         "//vendor/github.com/coreos/go-systemd/daemon:go_default_library",
         "//vendor/github.com/emicklei/go-restful:go_default_library",

Run ./hack/update-bazel.sh
```

**Comment by KielChan** [comment]
> ./hack/update-bazel.sh

Fine, I will check it.

**Comment by KielChan** [comment]
@SergeyKanzhelev    Thanks for your review. If all tests passed or still failed, I will re-write another PR to converge commit.

**Pull request closed without merging**